BlueprintUI/Sources/Layout/GridRow.swift

```diff
@@ -0,0 +1,303 @@
+            let yVector = _yVector(rowHeight: rowHeight)
+
+            // Compute frames.
+            var offset: CGFloat = 0
+            return sizes.map { size in
+                defer { offset += size.width + spacing }
+                let (yOrigin, height) = yVector(size.height)
+                return CGRect(x: offset, y: yOrigin, width: size.width, height: height)
+            }
+        }
+
+        /// When the layout width is defined, allot the children a portion of the available space.
+        ///
+        /// Determine the scale of a portion (points per portion), then use it to apply a width to each child.
+        func constrainedProportionalItemSizes(
+            in constraint: SizeConstraint,
+            availableWidth: CGFloat,
+            items: [IndexedProportionalItem]
+        ) -> [IndexedSize] {
+            guard items.count > 0 else {
+                return []
+            }
+
+            let portionSum = items.map { $0.proportion }.reduce(0, +)
+            precondition(portionSum > 0, "Proportions of a GridRow must sum to a positive number. Found sum: \(portionSum).")
+            let scale = availableWidth / portionSum
+
+            return items.map { (index, proportion, content) in
+                let width = scale * proportion
+                let fixedWidthConstraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                let size = CGSize(width: width, height: content.measure(in: fixedWidthConstraint).height)
+                return (index, size)
+            }
+        }
+
+        /// When the layout width is undefined, size children relative to each other.
+        ///
+        /// Measure each child's width to find the maximum scale (points per portion). The maximum scale
+        /// can be applied to each child to provide widths which satisfy the proportional requirements
+        /// between children and provide enough space to layout each child.
+        ///
+        /// ## Example:
+        ///      - proportions:      1,     2,     3
+        ///      - measured widths:  10,    25,    24
+        ///      - scales:           10,    12.5,  8
+        ///      - max scale:        12.5
+        ///      - widths:           12.5,  25,    37.5
+        private func unconstrainedProportionalItemSizes(
+            in constraint: SizeConstraint,
+            items: [IndexedProportionalItem]
+        ) -> [IndexedSize] {
+            var scale: CGFloat = 0
+            items.forEach { (_, proportion, content) in
+                let width = content.measure(in: constraint).width
+                scale = max(scale, width / proportion)
+            }
+
+            return items.map { (index, proportion, content) in
+                let width = scale * proportion
+                let constraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                let size = CGSize(width: width, height: content.measure(in: constraint).height)
```

Comment:
Can we skip this second measurement and use the height from the first?

Reply:
The overloading of the `constraint` here makes this a bit less legible, so I intend to fix it.

The first measurement measures the element against an unconstrained constraint. The second measurement occurs against a width-constrained constrained.

Because the height of some elements might grow when the width is constrained, I think we need this second measurement. Please correct me if I've misunderstood!

Reply:
> The first measurement occurs against an unconstrained constraint. The second measurement occurs against a width-constrained constrained.
> 
> Because the height of some elements might grow when the width is constrained, I think we need this second measurement. Please correct me if I've misunderstood!

If I'm understanding right, the width constraint used for the second measurement is guaranteed to be at least as wide as the size that the element reported in the first measurement. So while an element _could_ report a different size under that constraint, I think it would be unusual. Do you have a use case in mind where this would return a different size? I'm not sure it's a behavior we need to support.

(Having just done a performance pass I'm a bit wary of measurement calls.)

Reply:
The use case I had in mind was a label. 

Given an unconstrained width (the first measurement), it would have a height as tall as a single line. 

Given a constrained width (the second measurement), the height will in some cases be taller.

> the width constraint used for the second measurement is guaranteed to be at least as wide as the size that the element reported in the first measurement

I think that is correct, but I'm not sure it accounts for the height concern.

I also want to avoid unnecessary measurement and did try to reason through each call to see if it could be skipped. Still, I'm a novice, and my naivety is worth testing. Please let me know if you think my case is flawed.

Reply:
This seems correct – if the width deviates at all from previous measurements, you will need to re-measure, because things like labels will re-flow.

Reply:
OK, here's an example with a label that might help.

```swift
let label = UILabel()
label.text = "Hello world"
```

Let's measure the label with no constraint. Here we're asking the label: given an infinite amount of space, what is the size you want to be?

```swift
let unconstrainedSize = label.sizeThatFits(.zero)
print(unconstrainedSize) // (85.5, 20.5)
```

Now let's take that size, and use it as a constraint:

```swift
let constrainedSize = label.sizeThatFits(unconstrainedSize)
print(constrainedSize) // (85.5, 20.5)
```

The label reports the same size. This is because the constraint is large enough that the label does not need to reflow. We know it is large enough, because the label has already told us that this is the size it wants to be. For any well-behaved element, this should hold true.

Reply:
Got it. Apologies for being a bit dense here, this makes sense. I'll open a PR to fix it.